src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx

```diff
@@ -0,0 +1,396 @@
+    this->Internal->MaxMissingFiducials = 0;
+  }
+
+  XML_READ_SCALAR_ATTRIBUTE_NONMEMBER_OPTIONAL(float, MaxMeanRegistrationErrorMm, this->Internal->MaxMeanRegistrationErrorMm, deviceConfig);
+  if (this->Internal->MaxMeanRegistrationErrorMm < 0.1 || this->Internal->MaxMeanRegistrationErrorMm > 5.0)
+  {
+    LOG_WARNING("Invalid maximum mean registration error provided. Must be between 0.1 and 5 mm inclusive. Maximum mean registration error has been set to its default value of 5.0mm.");
+    this->Internal->MaxMeanRegistrationErrorMm = 2.0;
+  }
+
+  XML_READ_SCALAR_ATTRIBUTE_NONMEMBER_OPTIONAL(int, ActiveMarkerPairingTimeSec, this->Internal->ActiveMarkerPairingTimeSec, deviceConfig);
+  if (this->Internal->ActiveMarkerPairingTimeSec > 60)
+  {
+    LOG_WARNING("Are you sure you want to leave active marker pairing enabled for more than 60 seconds?");
+  }
+
+  enum TRACKING_TYPE
+  {
+    ACTIVE,
+    PASSIVE
+  };
+
+  XML_FIND_NESTED_ELEMENT_REQUIRED(dataSourcesElement, deviceConfig, "DataSources");
+  for (int nestedElementIndex = 0; nestedElementIndex < dataSourcesElement->GetNumberOfNestedElements(); nestedElementIndex++)
+  {
+    vtkXMLDataElement* toolDataElement = dataSourcesElement->GetNestedElement(nestedElementIndex);
+    if (STRCASECMP(toolDataElement->GetName(), "DataSource") != 0)
+    {
+      // if this is not a data source element, skip it
+      continue;
+    }
+    if (toolDataElement->GetAttribute("Type") != NULL && STRCASECMP(toolDataElement->GetAttribute("Type"), "Tool") != 0)
+    {
+      // if this is not a Tool element, skip it
+      continue;
+    }
+    std::string toolId(toolDataElement->GetAttribute("Id"));
+    if (toolId.empty())
+    {
+    // tool doesn't have ID needed to generate transform
+    LOG_ERROR("Failed to initialize Atracsys tool: DataSource Id is missing.");
+    continue;
+    }
+
+    TRACKING_TYPE toolTrackingType;
+    XML_READ_ENUM2_ATTRIBUTE_NONMEMBER_OPTIONAL(TrackingType, toolTrackingType, toolDataElement, "ACTIVE", ACTIVE, "PASSIVE", PASSIVE);
+    if (toolTrackingType == ACTIVE)
+    {
+      // active tool, can be loaded directly into FtkGeometryIdMappedToToolId
+      int ftkGeometryId = -1;
+      XML_READ_SCALAR_ATTRIBUTE_NONMEMBER_OPTIONAL(int, GeometryId, ftkGeometryId, toolDataElement);
+      if (ftkGeometryId != -1)
+      {
+        std::pair<int, std::string> thisTool(ftkGeometryId, toolId);
+        this->Internal->FtkGeometryIdMappedToToolId.insert(thisTool);
+      }
+      else
+      {
+        LOG_ERROR("Active tool with Id " << toolId << " is missing a GeometryId");
+        return PLUS_FAIL;
+      }
+    }
+    else if (toolTrackingType == PASSIVE)
+    {
+      // passive tool, load into PlusIdMappedToGeometryFilename to have geometry loaded in InternalConnect
+      const char* geometryFile;
+      if ((geometryFile = toolDataElement->GetAttribute("GeometryFile")) != NULL)
+      {
+        std::pair<std::string, std::string> thisTool(toolId, geometryFile);
```

Comment:
why don't we use a map<string,string> instead?

Reply:
We do, I just create the std::pair to make adding to the map more intuitive

Reply:
std::pair is probably the most useless container. Since "first" and "second" members are not named, they don't make things very clear or self-explaining.